third_party/nvidia/hopper/lib/Transforms/WarpSpecialization/WSCodePartition.cpp

```diff
@@ -1802,14 +1855,29 @@ void insertAsyncComm(
         producerOps.insert(pcOp.first);
         consumerOps.insert(pcOp.second);
       }
-      consumerOps.insert(c->getDstOp());
-      consumerOps.insert(getUniqueActualConsumer(c->getDstOp()));
+      if (c->channelKind == DataChannelKind::SMEMPost) {
+        auto *cPost = static_cast<ChannelPost *>(c);
+        SmallVector<Operation *> dsts;
+        cPost->getDstOps(dsts);
+        for (auto *dst : dsts) {
+          consumerOps.insert(dst);
+          auto consumers = getActualConsumers(dst);
+          for (auto *t : consumers) {
+            consumerOps.insert(t);
+            actualConsumerOps.insert(t);
+          }
+        }
+      } else {
+        consumerOps.insert(c->getDstOp());
+        consumerOps.insert(getUniqueActualConsumer(c->getDstOp()));
+        actualConsumerOps.insert(getUniqueActualConsumer(c->getDstOp()));
+      }
     }
 
-    // Assuming all ops
-    auto getHeadOp = [&](DenseSet<Operation *> ops,
-                         Operation *firstOp) -> Operation * {
-      auto block = firstOp->getBlock();
+    // Assuming all ops are under the same block.
+    auto getFirstOpInBlock = [&](DenseSet<Operation *> ops) -> Operation * {
```

Comment:
This currently copies the set (here and below)

```suggestion
    auto getFirstOpInBlock = [&](const DenseSet<Operation *>& ops) -> Operation * {
```